File: backend/app/schemas/report.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel


class ReportPaymentMethodOption(BaseModel):
    id: UUID
    name: str
    code: str


class ReportUserOption(BaseModel):
    id: UUID
    full_name: str
    username: str


class ReportNamedAmount(BaseModel):
    label: str
    amount: Decimal


class SalesDaySummary(BaseModel):
    total_vendido: Decimal
    cantidad_ventas: int
    cantidad_ventas_completadas: int
    cantidad_ventas_anuladas: int
    ticket_promedio: Decimal
    total_bruto: Decimal
    total_anulado: Decimal
    total_descuentos: Decimal
    total_efectivo: Decimal
    total_tarjeta: Decimal
    total_transferencia: Decimal
    total_otros: Decimal
    otros_desglose: list[ReportNamedAmount]


class SalesDaySaleItem(BaseModel):
    id: UUID
    folio: str
    fecha_hora: datetime
    hora: str
    cajero: str
    metodo_pago: str
    total: Decimal
    estado: str


class SalesDayAvailableFilters(BaseModel):
    metodos_pago: list[ReportPaymentMethodOption]
    cajeros: list[ReportUserOption]
    estados: list[str]


class SalesDayReportRead(BaseModel):
    fecha: date
    generado_en: datetime
    periodo_desde: datetime | None = None
    periodo_hasta: datetime | None = None
    resumen: SalesDaySummary
    ventas: list[SalesDaySaleItem]
    filtros_disponibles: SalesDayAvailableFilters


class CashDailyMovementItem(BaseModel):
    id: UUID
    fecha_hora: datetime
    tipo: str
    monto: Decimal
    motivo: str
    observacion: str | None = None
    usuario: str


class CashDailySaleItem(BaseModel):
    id: UUID
    folio: str
    fecha_hora: datetime
    cajero: str
    metodo_pago: str
    total: Decimal
    estado: str


class CashDailySummary(BaseModel):
    ventas_registradas: int
    ventas_completadas: int
    ventas_anuladas: int
    total_bruto: Decimal
    total_anulado: Decimal
    total_descuentos: Decimal
    total_vendido: Decimal
    total_efectivo: Decimal
    total_tarjeta: Decimal
    total_transferencia: Decimal
    total_otros: Decimal
    otros_desglose: list[ReportNamedAmount]
    caja_inicial: Decimal
    ingresos_manuales: Decimal
    retiros_manuales: Decimal
    caja_esperada: Decimal
    caja_contada: Decimal | None = None
    diferencia: Decimal | None = None


class CashDailyReportRead(BaseModel):
    fecha: date
    generado_en: datetime
    session_id: UUID
    estado: str
    caja: str
    ubicacion_caja: str | None = None
    usuario: str
    apertura: datetime
    cierre: datetime | None = None
    resumen: CashDailySummary
    ventas: list[CashDailySaleItem]
    movimientos: list[CashDailyMovementItem]

File: frontend/src/features/reports/types/report.ts
import type { User } from "../../auth/types/auth";
import type { Product } from "../../products/types/product";

export type ReportPaymentMethodOption = {
  id: string;
  name: string;
  code: string;
};

export type ReportUserOption = {
  id: string;
  full_name: string;
  username: string;
};

export type ReportNamedAmount = {
  label: string;
  amount: string;
};

export type VentasDiaResumen = {
  total_vendido: string;
  cantidad_ventas: number;
  cantidad_ventas_completadas: number;
  cantidad_ventas_anuladas: number;
  ticket_promedio: string;
  total_bruto: string;
  total_anulado: string;
  total_descuentos: string;
  total_efectivo: string;
  total_tarjeta: string;
  total_transferencia: string;
  total_otros: string;
  otros_desglose: ReportNamedAmount[];
};

export type VentaDiaItem = {
  id: string;
  folio: string;
  fecha_hora: string;
  hora: string;
  cajero: string;
  metodo_pago: string;
  total: string;
  estado: string;
};

export type VentasDiaReport = {
  fecha: string;
  generado_en: string;
  periodo_desde?: string | null;
  periodo_hasta?: string | null;
  resumen: VentasDiaResumen;
  ventas: VentaDiaItem[];
  filtros_disponibles: {
    metodos_pago: ReportPaymentMethodOption[];
    cajeros: ReportUserOption[];
    estados: string[];
  };
};

export type CajaDiariaMovimiento = {
  id: string;
  fecha_hora: string;
  tipo: "income" | "withdrawal";
  monto: string;
  motivo: string;
  observacion?: string | null;
  usuario: string;
};

export type CajaDiariaVenta = {
  id: string;
  folio: string;
  fecha_hora: string;
  cajero: string;
  metodo_pago: string;
  total: string;
  estado: string;
};

export type CajaDiariaResumen = {
  ventas_registradas: number;
  ventas_completadas: number;
  ventas_anuladas: number;
  total_bruto: string;
  total_anulado: string;
  total_descuentos: string;
  total_vendido: string;
  total_efectivo: string;
  total_tarjeta: string;
  total_transferencia: string;
  total_otros: string;
  otros_desglose: ReportNamedAmount[];
  caja_inicial: string;
  ingresos_manuales: string;
  retiros_manuales: string;
  caja_esperada: string;
  caja_contada?: string | null;
  diferencia?: string | null;
};

export type CajaDiariaReport = {
  fecha: string;
  generado_en: string;
  session_id: string;
  estado: string;
  caja: string;
  ubicacion_caja?: string | null;
  usuario: string;
  apertura: string;
  cierre?: string | null;
  resumen: CajaDiariaResumen;
  ventas: CajaDiariaVenta[];
  movimientos: CajaDiariaMovimiento[];
};

export type VentasDiaFilters = {
  search?: string;
  payment_method_id?: string;
  user_id?: string;
  status?: string;
};

export type CajaDiariaFilters = {
  session_id?: string;
  user_id?: string;
  report_date?: string;
};

export type SaleDetailReport = {
  id: string;
  user_id: string;
  cash_session_id?: string | null;
  sale_number: string;
  status: string;
  subtotal: string;
  discount_amount: string;
  tax_amount: string;
  total: string;
  notes?: string | null;
  cancellation_reason?: string | null;
  cancelled_at?: string | null;
  created_at: string;
  user: User;
  payments: Array<{
    id: string;
    payment_method_id: string;
    amount: string;
    reference?: string | null;
    payment_method: {
      id: string;
      name: string;
      code: string;
    };
  }>;
  details: Array<{
    id: string;
    product_id: string;
    quantity: string;
    unit_price: string;
    line_total: string;
    discount_amount: string;
    tax_amount: string;
    product: Product;
  }>;
};
